### PR TITLE
Refactor membership tests

### PR DESCRIPTION
**Pull request by mottosso**
Repair #234, at the expense of backwards incompatibility.

So as it turns out, QMetaType is gone (as [Fredrik pointed out a few days ago](https://github.com/mottosso/Qt.py/pull/207#issuecomment-323256841)) which means it can't be included per default by Qt.py anymore. They've essentially broken backwards compatibility for themselves, and in turn us as we depend on their implementation.

As for QtOpenGL (which I had to take out), it was only ever included for convenience and demo of Optional Submodules (#185) where modules were looked after, but included only if they existed. Rather than throwing an import error.

It simplifies things (by removing the separate membership test) and remains true to PySide2, at the cost of breaking backwards compatibility.

Thoughts? @MHendricks this affects your use of QMetaType and the way you'd have to work around it (if we merge this) is by leveraging [QtSiteConfig](https://github.com/mottosso/Qt.py#qtsiteconfigpy) to add it in yourself.

### CODE REVIEW

**Comment by fredrikaverpil** [comment]
@mottosso Don't we still need a script to identify common members across all bindings and output that (for inclusion in Qt.py's `_common_members`)?

**Comment by fredrikaverpil** [comment]
It seems the backwards breaking change may not have been intentional: https://gitter.im/PySide/pyside2?at=59a53033bc46472974d3410a

**Comment by fredrikaverpil** [comment]
I'm starting to wonder if `QtCore.QMetaType` ever existed in PySide2...

```bash
$ docker run -ti --rm --entrypoint bash fredrikaverpil/qt.py:2017
$ python2.7
Python 2.7.11 (default, Aug 15 2017, 19:33:56)
[GCC 4.8.3] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from PySide2 import QtCore
>>> QtCore.QMetaType
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'QMetaType'
```

**Comment by mottosso** [comment]
> How is it that this container manages to pass our current memberships test, where QtCore.QMetaType is included?!

Because it wasn't testing against `_common_members`, it was testing against the list built by the membership test module.

QMetaType can be included in `_common_members` and not trigger an error on `import Qt` for the same reason we can include `QtOpenGL`; it's optional. It's made available if it's available in the binding, such as PyQt5.

That's wrong though, we shouldn't let members through unless they have some chance of existing across all bindings. And it looks like QMetaType is not one of them.

**Comment by fredrikaverpil** [comment]
I noticed you left the membership test in the `entrypoint.sh` and I just removed it.

**Comment by mottosso** [comment]
> Don't we still need a script to identify common members across all bindings

Yeah, we could still have that. But as a separate process perhaps? We could use it as a tool to update `_common_members` from time to time.

**Comment by fredrikaverpil** [comment]
> Because it wasn't testing against _common_members, it was testing against the list built by the membership test module.

But how come tests for VFX Platform 2018 started complaining about QtCore.QMetaType all of a sudden?

**Comment by fredrikaverpil** [comment]
> Yeah, we could still have that. But as a separate process perhaps? We could use it as a tool to update _common_members from time to time.

That's exactly what I was thinking of. Just a helper tool to populate `_common_members`.

**Comment by mottosso** [comment]
> But how come tests for VFX Platform 2018 started complaining about QtCore.QMetaType all of a sudden?

[It's complaining](https://travis-ci.org/mottosso/Qt.py/jobs/268748958#L628) because it found members in PySide2 that it couldn't find elsewhere, such as QtQml. My guess is that either PyQt5 was compiled without those in this Docker image, or PySide2 didn't include it in previous Docker images.

What we can do is either expand on that exclusion list I made at the bottom of the membership test module, or do what we do in #240 and test on `_common_members` instead.

**Comment by fredrikaverpil** [comment]
I think we should merge this.

**Comment by fredrikaverpil** [comment]
Perhaps add a tool to identify the common members first. It could run as part of the tests or independently. But in either case, it could be used with the latest docker container.

**Comment by fredrikaverpil** [comment]
We could serially have each binding spit out its members to e.g. a json file. As a final step, the script collect the common members from the json files and output them in a separate json file. This could then be copy-pasted to be used in Qt.py.

**Comment by fredrikaverpil** [comment]
Hm, that sucks. PyQt4/5 doesn't have `__all__`...

**Comment by fredrikaverpil** [comment]
Ah, this was nice: https://docs.python.org/2/library/pkgutil.html

**Comment by fredrikaverpil** [comment]
Hm. Not sure what we could do to parse down into e.g. `QtWidgets`... @mottosso any ideas?

```
$ chmod +x run_membership.sh
$ ./run_membership.sh
PySide.QtCore
PySide.QtDeclarative
PySide.QtGui
PySide.QtHelp
PySide.QtMultimedia
PySide.QtNetwork
PySide.QtOpenGL
PySide.QtScript
PySide.QtScriptTools
PySide.QtSql
PySide.QtSvg
PySide.QtTest
PySide.QtUiTools
PySide.QtWebKit
PySide.QtXml
PySide.QtXmlPatterns
PySide._utils
PySide.phonon
PySide.scripts
PySide.scripts.uic
PySide.shiboken
PySide2.QtConcurrent
PySide2.QtCore
PySide2.QtGui
PySide2.QtHelp
PySide2.QtNetwork
PySide2.QtPrintSupport
PySide2.QtQml
PySide2.QtQuick
PySide2.QtQuickWidgets
PySide2.QtScript
PySide2.QtScriptTools
PySide2.QtSql
PySide2.QtSvg
PySide2.QtTest
PySide2.QtUiTools
PySide2.QtWebChannel
PySide2.QtWebKit
PySide2.QtWebKitWidgets
PySide2.QtWebSockets
PySide2.QtWidgets
PySide2.QtX11Extras
PySide2.QtXml
PySide2.QtXmlPatterns
PySide2._utils
PySide2.scripts
PySide2.scripts.uic
PySide2.shiboken2
PyQt4.Qt
PyQt4.QtCore
PyQt4.QtDeclarative
PyQt4.QtDesigner
PyQt4.QtGui
PyQt4.QtHelp
PyQt4.QtMultimedia
PyQt4.QtNetwork
PyQt4.QtOpenGL
PyQt4.QtScript
PyQt4.QtScriptTools
PyQt4.QtSql
PyQt4.QtSvg
PyQt4.QtTest
PyQt4.QtWebKit
PyQt4.QtXml
PyQt4.QtXmlPatterns
PyQt4.phonon
PyQt4.pyqtconfig
PyQt4.uic
PyQt4.uic.Compiler
PyQt4.uic.Compiler.compiler
PyQt4.uic.Compiler.indenter
PyQt4.uic.Compiler.misc
PyQt4.uic.Compiler.proxy_metaclass
PyQt4.uic.Compiler.qobjectcreator
PyQt4.uic.Compiler.qtproxies
PyQt4.uic.Loader
PyQt4.uic.Loader.loader
PyQt4.uic.Loader.qobjectcreator
PyQt4.uic.driver
PyQt4.uic.exceptions
PyQt4.uic.icon_cache
PyQt4.uic.objcreator
PyQt4.uic.port_v2
PyQt4.uic.port_v2.as_string
PyQt4.uic.port_v2.ascii_upper
PyQt4.uic.port_v2.invoke
PyQt4.uic.port_v2.load_plugin
PyQt4.uic.port_v2.proxy_base
PyQt4.uic.port_v2.string_io
PyQt4.uic.port_v3
PyQt4.uic.port_v3.as_string
PyQt4.uic.port_v3.ascii_upper
PyQt4.uic.port_v3.invoke
PyQt4.uic.port_v3.load_plugin
PyQt4.uic.port_v3.proxy_base
PyQt4.uic.port_v3.string_io
PyQt4.uic.properties
PyQt4.uic.pyuic
PyQt4.uic.uiparser
PyQt5.Enginio
PyQt5.Qt
PyQt5.QtBluetooth
PyQt5.QtCore
PyQt5.QtDBus
PyQt5.QtDesigner
PyQt5.QtGui
PyQt5.QtHelp
PyQt5.QtLocation
PyQt5.QtMultimedia
PyQt5.QtMultimediaWidgets
PyQt5.QtNetwork
PyQt5.QtNfc
PyQt5.QtOpenGL
PyQt5.QtPositioning
PyQt5.QtPrintSupport
PyQt5.QtQml
PyQt5.QtQuick
PyQt5.QtQuickWidgets
PyQt5.QtSensors
PyQt5.QtSerialPort
PyQt5.QtSql
PyQt5.QtSvg
PyQt5.QtTest
PyQt5.QtWebChannel
PyQt5.QtWebKit
PyQt5.QtWebKitWidgets
PyQt5.QtWebSockets
PyQt5.QtWidgets
PyQt5.QtX11Extras
PyQt5.QtXml
PyQt5.QtXmlPatterns
PyQt5._QOpenGLFunctions_2_0
PyQt5._QOpenGLFunctions_2_1
PyQt5._QOpenGLFunctions_4_1_Core
PyQt5.uic
PyQt5.uic.Compiler
PyQt5.uic.Compiler.compiler
PyQt5.uic.Compiler.indenter
PyQt5.uic.Compiler.misc
PyQt5.uic.Compiler.proxy_metaclass
PyQt5.uic.Compiler.qobjectcreator
PyQt5.uic.Compiler.qtproxies
PyQt5.uic.Loader
PyQt5.uic.Loader.loader
PyQt5.uic.Loader.qobjectcreator
PyQt5.uic.driver
PyQt5.uic.exceptions
PyQt5.uic.icon_cache
PyQt5.uic.objcreator
PyQt5.uic.port_v2
PyQt5.uic.port_v2.as_string
PyQt5.uic.port_v2.ascii_upper
PyQt5.uic.port_v2.invoke
PyQt5.uic.port_v2.load_plugin
PyQt5.uic.port_v2.proxy_base
PyQt5.uic.port_v2.string_io
PyQt5.uic.port_v3
PyQt5.uic.port_v3.as_string
PyQt5.uic.port_v3.ascii_upper
PyQt5.uic.port_v3.invoke
PyQt5.uic.port_v3.load_plugin
PyQt5.uic.port_v3.proxy_base
PyQt5.uic.port_v3.string_io
PyQt5.uic.properties
PyQt5.uic.pyuic
PyQt5.uic.uiparser
```

**Comment by fredrikaverpil** [comment]
Ok, I think I got it now.

```
$ ./run_membership.sh
...
PySide.QtGui.QPrinter
PySide.QtGui.QPrinterInfo
PySide.QtGui.QProgressBar
PySide.QtGui.QProgressDialog
PySide.QtGui.QProxyModel
PySide.QtGui.QPushButton
PySide.QtGui.QPyTextObject
...
```

The list is enormous. Now we just need that on JSON form and compare.

File: membership.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+                onerror=lambda x: None):
+            MODULES.append(modname)
+            base = modname[:modname.rfind('.')]
+            mod = modname[modname.rfind('.')+1:]
+            try:
+                exec('from ' + base + ' import ' + mod)
+                exec('raw_members = dir(' + mod + ')')
+
+                members = []
+                for member in raw_members:
```
[the remark this answers is not in the document]
Comment (fredrikaverpil) [reply membership.py]: Good boy! 🍖

**Comment by fredrikaverpil** [comment]
Turns out I did hit a couple of issues due to e.g. `the PyQt5.QtCore and PyQt4.QtCore modules both wrap the QObject class`... so now each binding analysis is performed separately.

I also added JSON export, so now we'll get four JSON files; `PySide.json`, `PySide2.json`, `PyQt4.json` and `PyQt5.json`.

**Comment by fredrikaverpil** [comment]
Okay, now I export one JSON file per binding and then those are compared and a `common_members.json` is spit out. But, ok, I'm dumb. Because this doesn't take into account that e.g. `QtWidgets` doesn't exist on PyQt4/PySide (as it's called `QtGui` there etc)... so the comparison is quite useless...

Sigh. Dead end then. How did I not see this coming?!
I'm going to sleep now.

**Comment by mottosso** [comment]
@fredrikaverpil Should I merge this?

**Comment by fredrikaverpil** [comment]
@mottosso I think we should merge your original PR, yes.

The work I added to his PR, however, should not be merged.

I added a script (`run_membership.py`) which will output `PySide.json`, `PyQt4.json`, `PySide2.json`, `PyQt5.json`. In each one, all of the binding's members are listed. Then, a comparison was made which produced a `common_membership.json` which would include the shared members throughout all of those JSON files. But what I didn't think of is that e.g. PySide has its widgets under `QtGui` and PySide2 has its widgets under QtWidgets. So the resulting `common_membership.json` is useless to us.

I'm not sure how we can make a script which properly helps us generate the membership dictionary.

**Comment by fredrikaverpil** [comment]
I mean, I'm not sure how we can create a script which will help us identify common members without having to maintain the different changes between Qt4/Qt5 in such a script.

**Comment by fredrikaverpil** [comment]
Hm... perhaps it's not too much to maintain, actually.

Since we do this for PySide2/PyQt4:

```python
    if hasattr(Qt, "_QtGui"):
        setattr(Qt, "QtWidgets", _new_module("QtWidgets"))
        setattr(Qt, "_QtWidgets", Qt._QtGui)
```

...I should just be able to force the JSON comparison script to compare QtWidgets against QtGui for those bindings. And everything should actually work.

**Comment by fredrikaverpil** [comment]
Holy crap, I think it worked.

```json
{
    "QtCore": [
        "QAbstractTransition",
        "QTextBoundaryFinder",
        "QTimerEvent",
        "QTextDecoder",
        "QtSystemMsg",
        "QXmlStreamAttributes",
        "QPointF",
        "QRunnable",
        "QMargins",
        "QPluginLoader",
        "QSizeF",
        "QLineF",
        "QStateMachine",
        "QTimeLine",
        "QtDebugMsg",
        "QThread",
        "QSettings",
        "QAbstractState",
        "Qt",
        "QXmlStreamAttribute",
        "QMetaObject",
        "QMetaMethod",
        "QBasicTimer",
        "QT_TR_NOOP_UTF8",
        "QWaitCondition",
        "QUrl",
        "QEvent",
        "QtCriticalMsg",
        "QMetaEnum",
        "QRegExp",
        "qsrand",
        "QTime",
        "QIODevice",
        "QDir",
        "QRectF",
        "QAbstractListModel",
        "QXmlStreamNotationDeclaration",
        "QAnimationGroup",
        "QHistoryState",
        "QXmlStreamReader",
        "QSysInfo",
        "QFileInfo",
        "QFile",
        "QEasingCurve",
        "qDebug",
        "qVersion",
        "QThreadPool",
        "QTimer",
        "QXmlStreamEntityResolver",
        "QXmlStreamNamespaceDeclaration",
        "QByteArray",
        "QLine",
        "QAbstractEventDispatcher",
        "QParallelAnimationGroup",
        "QSequentialAnimationGroup",
        "QXmlStreamWriter",
        "QTextCodec",
        "QXmlStreamEntityDeclaration",
        "QT_TRANSLATE_NOOP",
        "qWarning",
        "QFinalState",
        "QtWarningMsg",
        "QtFatalMsg",
        "QTranslator",
        "QT_TR_NOOP",
        "qChecksum",
        "QSystemSemaphore",
        "QProcessEnvironment",
        "QElapsedTimer",
        "QProcess",
        "QDirIterator",
        "qUnregisterResourceData",
        "QMetaProperty",
        "QPropertyAnimation",
        "QSignalTransition",
        "QSize",
        "QDateTime",
        "QTextEncoder",
        "qRegisterResourceData",
        "QPauseAnimation",
        "QVariantAnimation",
        "qrand",
        "QReadLocker",
        "QLibraryInfo",
        "QFileSystemWatcher",
        "QTextStreamManipulator",
        "QResource",
        "qFatal",
        "qFuzzyCompare",
        "QSocketNotifier",
        "QBuffer",
        "QDate",
        "QCoreApplication",
        "QMutexLocker",
        "QDataStream",
        "QCryptographicHash",
        "qCritical",
        "qIsNaN",
        "QAbstractItemModel",
        "QRect",
        "QPoint",
        "QMimeData",
        "QReadWriteLock",
        "qIsFinite",
        "QAbstractTableModel",
        "QtMsgType",
        "qIsNull",
        "QTextStream",
        "QAbstractAnimation",
        "QByteArrayMatcher",
        "QEventTransition",
        "QTemporaryFile",
        "QChildEvent",
        "QMetaClassInfo",
        "qIsInf",
        "QPersistentModelIndex",
        "QSemaphore",
        "QObject",
        "QWriteLocker",
        "QLocale",
        "QMutex",
        "QState",
        "QModelIndex",
        "qAbs",
        "QSignalMapper",
        "QBitArray",
        "QDynamicPropertyChangeEvent",
        "QEventLoop",
        "QGenericArgument",
        "qAddPostRoutine",
        "QGenericReturnArgument"
    ],
    "QtNetwork": [
        "QNetworkProxy",
        "QNetworkConfigurationManager",
        "QNetworkCookie",
        "QAbstractSocket",
        "QNetworkAddressEntry",
        "QHostInfo",
        "QTcpSocket",
        "QNetworkDiskCache",
        "QNetworkProxyFactory",
        "QAuthenticator",
        "QLocalServer",
        "QNetworkCacheMetaData",
        "QNetworkReply",
        "QLocalSocket",
        "QNetworkConfiguration",
        "QAbstractNetworkCache",
        "QHostAddress",
        "QSsl",
        "QNetworkSession",
        "QNetworkCookieJar",
        "QNetworkInterface",
        "QNetworkProxyQuery",
        "QTcpServer",
        "QNetworkRequest",
        "QNetworkAccessManager",
        "QUdpSocket"
    ],
    "QtSql": [
        "QSqlField",
        "QSqlTableModel",
        "QSqlResult",
        "QSqlRelationalTableModel",
        "QSqlDatabase",
        "QSql",
        "QSqlIndex",
        "QSqlQueryModel",
        "QSqlRelationalDelegate",
        "QSqlQuery",
        "QSqlRecord",
        "QSqlError",
        "QSqlDriver",
        "QSqlRelation",
        "QSqlDriverCreatorBase"
    ],
    "QtSvg": [
        "QGraphicsSvgItem",
        "QSvgRenderer",
        "QSvgGenerator",
        "QSvgWidget"
    ],
    "QtHelp": [
        "QHelpEngineCore",
        "QHelpSearchResultWidget",
        "QHelpContentModel",
        "QHelpSearchQuery",
        "QHelpIndexWidget",
        "QHelpContentWidget",
        "QHelpSearchQueryWidget",
        "QHelpIndexModel",
        "QHelpSearchEngine",
        "QHelpContentItem",
        "QHelpEngine"
    ],
    "QtWebKit": [],
    "QtGui": [
        "QFocusEvent",
        "QTextLayout",
        "QTextObject",
        "QPaintEngine",
        "QTextOption",
        "QMatrix2x2",
        "QTextDocumentFragment",
        "QActionEvent",
        "qAlpha",
        "qRgb",
        "QIconDragEvent",
        "QPainterPathStroker",
        "QTextLine",
        "QWindowStateChangeEvent",
        "QTextBlock",
        "QKeySequence",
        "QDragLeaveEvent",
        "QMatrix2x4",
        "QHelpEvent",
        "QMovie",
        "QBrush",
        "QPolygonF",
        "QTextLength",
        "QTextItem",
        "QDragEnterEvent",
        "QTouchEvent",
        "QColor",
        "QDragMoveEvent",
        "QTextBlockUserData",
        "qGray",
        "QImageWriter",
        "QMatrix4x3",
        "QLinearGradient",
        "QPalette",
        "QContextMenuEvent",
        "QTextListFormat",
        "QPen",
        "QTextFrameFormat",
        "QShowEvent",
        "QTextBlockFormat",
        "QResizeEvent",
        "QInputEvent",
        "QStandardItemModel",
        "QTextTableFormat",
        "QHoverEvent",
        "QPicture",
        "QPixmap",
        "QTextList",
        "QRegion",
        "QPaintEngineState",
        "QMouseEvent",
        "QBitmap",
        "QTextCharFormat",
        "QTextTableCell",
        "QMatrix2x3",
        "QFontInfo",
        "qGreen",
        "QHideEvent",
        "QTextFrame",
        "QTextImageFormat",
        "QPaintEvent",
        "QPainterPath",
        "QVector4D",
        "QTextObjectInterface",
        "QIconEngine",
        "QIntValidator",
        "QMatrix3x4",
        "qRed",
        "QAbstractTextDocumentLayout",
        "QFontDatabase",
        "QInputMethodEvent",
        "qRgba",
        "QTextTable",
        "qBlue",
        "QTextTableCellFormat",
        "QVector2D",
        "QMoveEvent",
        "QCloseEvent",
        "QVector3D",
        "QStatusTipEvent",
        "QKeyEvent",
        "QConicalGradient",
        "QFont",
        "QDrag",
        "QPictureIO",
        "QPolygon",
        "QTextFragment",
        "QMatrix3x3",
        "QDoubleValidator",
        "QRegExpValidator",
        "qIsGray",
        "QImageReader",
        "QTextBlockGroup",
        "QQuaternion",
        "QFontMetricsF",
        "QDropEvent",
        "QIcon",
        "QSyntaxHighlighter",
        "QValidator",
        "QPainter",
        "QTextFormat",
        "QWheelEvent",
        "QCursor",
        "QTextCursor",
        "QMatrix4x2",
        "QClipboard",
        "QMatrix3x2",
        "QGradient",
        "QRadialGradient",
        "QTextInlineObject",
        "QTabletEvent",
        "QStandardItem",
        "QImage",
        "QWhatsThisClickedEvent",
        "QImageIOHandler",
        "QTransform",
        "QSessionManager",
        "QFileOpenEvent",
        "QPaintDevice",
        "QPixmapCache",
        "QFontMetrics",
        "QTextDocument",
        "QMatrix4x4",
        "QShortcutEvent"
    ],
    "QtXmlPatterns": [
        "QXmlNodeModelIndex",
        "QXmlFormatter",
        "QXmlSchema",
        "QXmlName",
        "QAbstractXmlNodeModel",
        "QXmlNamePool",
        "QXmlResultItems",
        "QSourceLocation",
        "QAbstractUriResolver",
        "QAbstractXmlReceiver",
        "QXmlQuery",
        "QXmlItem",
        "QXmlSerializer",
        "QAbstractMessageHandler",
        "QXmlSchemaValidator"
    ],
    "QtWidgets": [
        "QGraphicsRectItem",
        "QGraphicsTransform",
        "QAbstractItemDelegate",
        "QGraphicsEffect",
        "QStyleOptionButton",
        "QItemEditorCreatorBase",
        "QErrorMessage",
        "QStackedLayout",
        "QApplication",
        "QPushButton",
        "QDoubleSpinBox",
        "QToolButton",
        "QGraphicsOpacityEffect",
        "QToolBar",
        "QStyleOptionGraphicsItem",
        "QCommonStyle",
        "QTabWidget",
        "QMenu",
        "QTableWidgetSelectionRange",
        "QFrame",
        "QGraphicsWidget",
        "QListView",
        "QTreeWidgetItemIterator",
        "QGraphicsLineItem",
        "QGraphicsDropShadowEffect",
        "QTapGesture",
        "QCalendarWidget",
        "QDesktopWidget",
        "QStyleOptionSizeGrip",
        "QGraphicsSceneHoverEvent",
        "QItemDelegate",
        "QGraphicsLinearLayout",
        "QAction",
        "QGraphicsItemGroup",
        "QSplitterHandle",
        "QAbstractGraphicsShapeItem",
        "QActionGroup",
        "QMenuBar",
        "QUndoGroup",
        "QUndoCommand",
        "QScrollBar",
        "QHeaderView",
        "QSplitter",
        "QStyleOptionTitleBar",
        "QTextEdit",
        "QPinchGesture",
        "QPlainTextEdit",
        "QGraphicsView",
        "QStyleOptionGroupBox",
        "QWidget",
        "QGraphicsLayout",
        "QStyleOptionFocusRect",
        "QWidgetItem",
        "QMdiSubWindow",
        "QWizardPage",
        "QItemEditorFactory",
        "QDialogButtonBox",
        "QStyleOptionToolButton",
        "QFocusFrame",
        "QMdiArea",
        "QTreeView",
        "QGraphicsScene",
        "QWidgetAction",
        "QGraphicsAnchorLayout",
        "QAbstractButton",
        "QGraphicsAnchor",
        "QGraphicsSimpleTextItem",
        "QStyleOptionViewItem",
        "QGraphicsScale",
        "QCheckBox",
        "QGraphicsRotation",
        "QTabBar",
        "QPlainTextDocumentLayout",
        "QLabel",
        "QTableWidgetItem",
        "QStyleOptionComplex",
        "QComboBox",
        "QGraphicsPixmapItem",
        "QGraphicsSceneEvent",
        "QGestureEvent",
        "QStyleOptionProgressBar",
        "QToolBox",
        "QStyleOption",
        "QStyleOptionDockWidget",
        "QStyleOptionTabBarBase",
        "QGraphicsEllipseItem",
        "QMainWindow",
        "QInputDialog",
        "QGraphicsSceneResizeEvent",
        "QStyleOptionTab",
        "QGraphicsSceneMoveEvent",
        "QGraphicsSceneMouseEvent",
        "QTreeWidget",
        "QFileSystemModel",
        "QGraphicsSceneDragDropEvent",
        "QLCDNumber",
        "QDateEdit",
        "QStyleOptionRubberBand",
        "QGraphicsBlurEffect",
        "QFormLayout",
        "QDockWidget",
        "QProgressDialog",
        "QPanGesture",
        "QTreeWidgetItem",
        "QStyleOptionTabWidgetFrame",
        "QLineEdit",
        "QStyleOptionComboBox",
        "QScrollArea",
        "QFileIconProvider",
        "QStyleOptionSlider",
        "QGestureRecognizer",
        "qApp",
        "QSystemTrayIcon",
        "QStackedWidget",
        "QBoxLayout",
        "QTapAndHoldGesture",
        "QUndoView",
        "QAbstractSlider",
        "QSpinBox",
        "QGridLayout",
        "QMessageBox",
        "QGraphicsTextItem",
        "QToolTip",
        "QAbstractItemView",
        "QListWidget",
        "QWhatsThis",
        "QMouseEventTransition",
        "QTableView",
        "QLayoutItem",
        "QFontComboBox",
        "QStyleHintReturnVariant",
        "QSizeGrip",
        "QGraphicsProxyWidget",
        "QStyleFactory",
        "QGraphicsObject",
        "QButtonGroup",
        "QStylePainter",
        "QStyleHintReturnMask",
        "QStyleOptionHeader",
        "QFileDialog",
        "QGraphicsLayoutItem",
        "QGraphicsGridLayout",
        "QStyleOptionToolBox",
        "QStyleOptionMenuItem",
        "QGroupBox",
        "QAbstractSpinBox",
        "QDialog",
        "QSizePolicy",
        "QStyleOptionFrame",
        "QRubberBand",
        "QStatusBar",
        "QWizard",
        "QShortcut",
        "QKeyEventTransition",
        "QDateTimeEdit",
        "QStyledItemDelegate",
        "QGraphicsSceneHelpEvent",
        "QColorDialog",
        "QDial",
        "QFontDialog",
        "QStyleHintReturn",
        "QLayout",
        "QHBoxLayout",
        "QGraphicsPathItem",
        "QGraphicsPolygonItem",
        "QStyle",
        "QStyleOptionToolBar",
        "QTimeEdit",
        "QCompleter",
        "QGraphicsSceneWheelEvent",
        "QRadioButton",
        "QGesture",
        "QSplashScreen",
        "QTextBrowser",
        "QColumnView",
        "QDirModel",
        "QListWidgetItem",
        "QCommandLinkButton",
        "QUndoStack",
        "QSpacerItem",
        "QGraphicsColorizeEffect",
        "QVBoxLayout",
        "QGraphicsSceneContextMenuEvent",
        "QProgressBar",
        "QDataWidgetMapper",
        "QStyleOptionSpinBox",
        "QSlider",
        "QSwipeGesture",
        "QGraphicsItem",
        "QTableWidget",
        "QAbstractScrollArea"
    ],
    "QtTest": [
        "QTest"
    ],
    "QtXml": [
        "QDomEntity",
        "QDomNamedNodeMap",
        "QXmlLocator",
        "QDomDocument",
        "QXmlLexicalHandler",
        "QDomProcessingInstruction",
        "QDomCharacterData",
        "QXmlDefaultHandler",
        "QDomEntityReference",
        "QXmlContentHandler",
        "QXmlInputSource",
        "QXmlReader",
        "QXmlNamespaceSupport",
        "QDomAttr",
        "QDomImplementation",
        "QDomDocumentType",
        "QDomDocumentFragment",
        "QXmlAttributes",
        "QXmlEntityResolver",
        "QDomText",
        "QDomNodeList",
        "QXmlDTDHandler",
        "QXmlParseException",
        "QXmlSimpleReader",
        "QDomElement",
        "QDomCDATASection",
        "QXmlErrorHandler",
        "QDomComment",
        "QDomNode",
        "QDomNotation",
        "QXmlDeclHandler"
    ]
}
```

**Comment by fredrikaverpil** [comment]
I quickly made the common members JSON sorted. Since the keys were not sorted in Qt.py the diff became a little difficult to read... I haven't had time to read it through yet. But at first glance, looks like this'll work!

**Comment by fredrikaverpil** [comment]
@mottosso I made it easier to spot the differences after having updated the Qt.py file's `_common_members` using the output from my comparison script. See here: https://github.com/mottosso/Qt.py/pull/240/commits/6198123ed497a55fdf02f2b4605e31e49660626a

After posting this message, I'll add the members back in, sorted. Which will make the overall diff a bit harder to read -- but will make future changes easier to read, as they are generated by the `membership.py` test the same way.

I think this is all actually really nice and think we should merge this as soon as possible, so that we can move on with other PRs like #234

The `membership.py` script could use some cleaning up to make code more grokable. But I have limited time on my hands and would prefer to do that down the line, if needed.

I don't think we need to edit the README.md for this change, as the fundamentals of Qt.py are the same (I'm reading what's under the ["Project goals"](https://github.com/mottosso/Qt.py#project-goals) headline.

**Comment by fredrikaverpil** [comment]
@mottosso this is now ready to be merged as far as I can tell.

**Comment by mottosso** [comment]
@fredrikaverpil Are you sure those new members exist across all bindings?

**Comment by fredrikaverpil** [comment]
I'm pretty sure, yes.

Each binding is inspected and the script attempts to import all modules and their submodules/members. All successful imports are exported to a `[binding_name].json` file. Then all JSON files are compared and if a member exists across all JSON files, it is added to `common_members.json`.

So, all of those members in the latest Qt.py update are imported on all bindings successfully. I exit the Python script for each processing of every binding to prevent issues (with e.g. SIP between PyQt4/PyQt5). I made the script a bit more verbose to show each import being performed successfully. Here's the log showing all successful imports as well as all warnings where an import could not be made: https://gist.github.com/fredrikaverpil/6fb1e6a2273fcd877b4b56c91738d36f

I would be surprised if this list is not accurate.

You can just run `./run_membership.sh` yourself and inspect the JSON files, but I uploaded the JSON files onto [a public Google Drive folder](https://drive.google.com/drive/u/0/folders/0B0b1IUnYqENgVW1vWEN3dlhPNEU). If you look inside the different JSONs, you'll see that some bindings (of course) have stuff which doesn't exist in the other JSON files, and therefore these members doesn't make it over onto `common_members.json`.

**Comment by fredrikaverpil** [comment]
Hm.. 

```
======================================================================
FAIL: Doctest: test_caveats.test_13_qtwidgets_qapp
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/python2.7/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/usr/local/python2.7/lib/python2.7/doctest.py", line 2226, in runTest
    raise self.failureException(self.format_failure(new.getvalue()))
AssertionError: Failed doctest test for test_13_qtwidgets_qapp
  File "/workdir/Qt.py/test_caveats.py", line 211, in test_13_qtwidgets_qapp
----------------------------------------------------------------------
File "/workdir/Qt.py/test_caveats.py", line 220, in test_13_qtwidgets_qapp
Failed example:
    "qApp" in dir(QtWidgets)
Expected:
    False
Got:
    True
```

**Comment by fredrikaverpil** [comment]
@mottosso sorry, didn't see your first post on qApp. I'll amend `membership.py` to exclude it.

**Comment by fredrikaverpil** [comment]
Ok, I bumped the version too. Should be fine to merge now.

File: membership.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+                                print('WARNING: Skipped import',
+                                      modname, error)
+                except (NameError) as error:
+                    print('WARNING: Skipped dir() command', modname, error)
+
+        # Remove duplicates and sort
+        MODULES = sorted(list(set(MODULES)))
+
+        # Print all modules (for debugging)
+        # for module in MODULES:
```
Comment (mottosso) [inline membership.py]: How about putting this and the `pprint` below under an `if` for [`QT_VERBOSE`](https://github.com/mottosso/Qt.py/blob/07f631545570c62310248e2a2545ea3103e811bc/Qt.py#L52)?

**Comment by mottosso** [comment]
Looks very slick! Happy to merge after that one change.

File: membership.py (hunk 2 above)
Comment (fredrikaverpil) [reply membership.py]: That's a nice idea. I'll make that change here now.

**Comment by mottosso** [comment]
Looking good, merge at will! Thanks @fredrikaverpil!

**Comment by fredrikaverpil** [comment]
Awesome. This change is kickass, to be honest.

**Pull request merged**